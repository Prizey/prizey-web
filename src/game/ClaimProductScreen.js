import React from 'react'
import { connect } from 'react-redux'
import { Link } from '@reach/router'
import { withStyles } from '@material-ui/core/styles'
import { Button, Typography } from '@material-ui/core'

import GoBack from 'design/GoBack/GoBack'
import Layout from 'design/Layout/Layout'
import Caption from 'design/Caption/Caption'

const styles = theme => ({
  icon: {
    height: theme.spacing.md,
  },
  productImage: {
    display: 'block',
    margin: 'auto',
    marginBottom: theme.spacing.md,
    marginTop: theme.spacing.md,
    width: '130px',
  },
  quantity: {
    marginLeft: theme.spacing.xs,
    marginRight: theme.spacing.xs,
    width: theme.spacing.sm,
  },
  rightIcon: {
    alignItems: 'center',
    display: 'flex',
    height: theme.spacing.md,
    position: 'absolute',
    right: 0,
  },
})

const InnerComponent = withStyles(styles)(
  ({ product, classes, difficulty }) => (
    <Layout
      leftIcon={<GoBack to={`game/${difficulty}`} />}
      caption={<Caption difficulty={difficulty} />}
    >
      <Typography align="center" variant="h5">
        {product.title}
      </Typography>

      <img
        alt={product.title}
        src={product.image}
        className={classes.productImage}
      />

      <Typography align="center">MSRP: ${product.price}</Typography>

      <Button
        variant="contained"
        color="secondary"
        fullWidth
        component={props => <Link to="/" {...props} />}
      >
        <span>SELL IT BACK</span>
        <div className={classes.rightIcon}>
          <img
            src="/icons/diamond.png"
            alt="diamond"
            className={classes.icon}
          />
          <span className={classes.quantity}>3</span>
        </div>
      </Button>
      <Button
        variant="contained"
        color="primary"
        fullWidth
        onClick={() => {
          // eslint-disable-next-line no-alert
          window.alert(`shipping process of the item ${product.title}`)
        }}
      >
        I WANT IT!
      </Button>
    </Layout>
  ),
)

export const mapState = ({ order = {} }) => ({
  product: order.product,
})

export default connect(
  mapState,
  null,
)(({ difficulty, product }) => (
  <InnerComponent product={product} difficulty={difficulty} />
))
