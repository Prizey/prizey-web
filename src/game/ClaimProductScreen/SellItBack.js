import React, { Fragment } from 'react'
import { withStyles } from '@material-ui/core/styles'
import { Button, Typography } from '@material-ui/core'
import CircularProgress from '@material-ui/core/CircularProgress'

import PlayAgain from 'design/PlayAgain'
import TransactionComponent from 'game/TransactionComponent'

const TICKET_AMOUNT = 3

const styles = theme => ({
  icon: {
    height: theme.spacing.md,
  },
  quantity: {
    marginLeft: theme.spacing.xs,
    marginRight: theme.spacing.xs,
    width: theme.spacing.sm,
  },
  rightIcon: {
    alignItems: 'center',
    display: 'flex',
    height: theme.spacing.md,
    position: 'absolute',
    right: 0,
  },
})

export const handleConfirm = ({ creating, create, amount }) => () => {
  if (!creating) {
    create({ amount })
  }
}

export class SellItBackComponent extends React.Component {
  state = {
    dialogIsOpen: false,
  }

  renderError = ({ error }) =>
    error && (
      <Typography color="error" align="center" variant="h6">
        {error}
      </Typography>
    )

  renderButton = ({ creating, classes }) => (
    <Button
      aria-label="Sell it Back"
      variant="contained"
      color="secondary"
      fullWidth
      onClick={() => this.setState({ dialogIsOpen: true })}
    >
      {creating ? (
        <CircularProgress color="inherit" size={30} />
      ) : (
        <Fragment>
          <span>SELL IT BACK</span>
          <div className={classes.rightIcon}>
            <img
              src="/icons/diamond.png"
              alt="diamond"
              className={classes.icon}
            />
            <span className={classes.quantity}>{TICKET_AMOUNT}</span>
          </div>
        </Fragment>
      )}
    </Button>
  )

  render() {
    const { dialogIsOpen } = this.state

    return (
      <Fragment>
        {this.renderError(this.props)}
        <PlayAgain
          aria-label="Play again modal"
          isOpen={dialogIsOpen}
          close={() => this.setState({ dialogIsOpen: false })}
          confirm={handleConfirm(this.props)}
        >
          You&apos;ll sell the product and get diamonds. This can&apos;t be
          undone.
        </PlayAgain>

        {this.renderButton(this.props)}
      </Fragment>
    )
  }
}

export const afterCreate = ({
  navigate,
  currentUser,
  setCurrentUser,
}) => () => {
  setCurrentUser({
    ...currentUser,
    tickets: currentUser.tickets + TICKET_AMOUNT,
  })
  navigate('/sold-back')
}

export default withStyles(styles)(props => (
  <TransactionComponent
    render={renderProps => <SellItBackComponent {...props} {...renderProps} />}
    afterCreate={afterCreate(props)}
  />
))
