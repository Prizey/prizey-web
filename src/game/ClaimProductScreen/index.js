import ClaimProductScreen from './ClaimProductScreen'

export default ClaimProductScreen
