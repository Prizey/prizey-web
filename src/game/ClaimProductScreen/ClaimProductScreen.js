import React from 'react'
import { connect } from 'react-redux'
import { Redirect } from '@reach/router'
import { Button, Typography } from '@material-ui/core'

import { clearProduct } from 'store/basket/actions'

import GoBack from 'design/GoBack/GoBack'
import Layout from 'design/Layout/Layout'
import Caption from 'design/Caption/Caption'
import ProductImage from 'design/ProductImage'

import SellItBack from './SellItBack'

const InnerComponent = ({ product, classes, difficulty, navigate, ...props }) =>
  product ? (
    <Layout
      leftIcon={<GoBack to="/" />}
      caption={<Caption difficulty={difficulty} />}
    >
      <Typography align="center" variant="h5">
        {product.title}
      </Typography>

      <ProductImage product={product} />
      <Typography align="center">MSRP: ${product.price}</Typography>

      <SellItBack
        amount={3}
        clearProduct={props.clearProduct}
        navigate={navigate}
      />

      <Button
        variant="contained"
        color="primary"
        fullWidth
        aria-label="I want it"
        onClick={() => {
          navigate('/shipping-info')
        }}
      >
        I WANT IT!
      </Button>
    </Layout>
  ) : (
    <Redirect to="/" noThrow />
  )

export const mapState = ({ basket = {} }) => ({
  product: basket.product,
})

const ClaimProductScreen = connect(
  mapState,
  { clearProduct },
)(props => <InnerComponent {...props} />)

export default ({ currentUser, ...props }) =>
  currentUser ? (
    <ClaimProductScreen {...props} />
  ) : (
    <Redirect to="/sign-in" noThrow />
  )
