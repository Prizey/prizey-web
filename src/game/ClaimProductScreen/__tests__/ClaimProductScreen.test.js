import React from 'react'
import renderer from 'react-test-renderer'
import ClaimProductScreen, { mapState } from '../ClaimProductScreen'

jest.mock('../../../design/Layout/RegisterPageView')
jest.mock('design/UserBalance', () => () => null)

jest.mock('react-redux', () => ({
  connect: () => Component => props => (
    <Component
      {...props}
      product={{ image: '/mock/trump-mask.png', title: 'Trump mask' }}
    />
  ),
}))

jest.mock('croods', () => ({
  New: props => (
    <div {...props}>
      New - <div>render - {props.render(props)}</div>
    </div>
  ),
}))

it('renders correctly', () => {
  const tree = renderer
    .create(<ClaimProductScreen currentUser difficulty="easy" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('redirects correctly', () => {
  const tree = renderer.create(<ClaimProductScreen />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('map the state to props', () => {
  const basket = {
    product: { image: '/mock/trump-mask.png', title: 'Trump mask' },
  }

  expect(mapState({ basket })).toEqual({ product: basket.product })
})

it('trigger the alert', () => {
  const navigate = jest.fn()
  const tree = renderer.create(
    <ClaimProductScreen difficulty="easy" currentUser navigate={navigate} />,
  ).root

  tree.findByProps({ 'aria-label': 'I want it' }).props.onClick()
  expect(navigate).toHaveBeenCalledTimes(1)
})
