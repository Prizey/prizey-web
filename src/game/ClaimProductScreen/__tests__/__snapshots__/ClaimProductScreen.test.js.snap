// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redirects correctly 1`] = `null`;

exports[`renders correctly 1`] = `
<div
  className="Component-root-8"
>
  <div
    className="Component-header-6"
  >
    <a
      aria-label="My profile"
      className="MuiButtonBase-root-16 MuiIconButton-root-10 Component-root-9"
      href="/profile"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="MuiIconButton-label-15"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-36 MuiSvgIcon-fontSizeLarge-44"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <g>
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM7.07 18.28c.43-.9 3.05-1.78 4.93-1.78s4.51.88 4.93 1.78C15.57 19.36 13.86 20 12 20s-3.57-.64-4.93-1.72zm11.29-1.45c-1.43-1.74-4.9-2.33-6.36-2.33s-4.93.59-6.36 2.33C4.62 15.49 4 13.82 4 12c0-4.41 3.59-8 8-8s8 3.59 8 8c0 1.82-.62 3.49-1.64 4.83z"
            />
            <path
              d="M12 6c-1.94 0-3.5 1.56-3.5 3.5S10.06 13 12 13s3.5-1.56 3.5-3.5S13.94 6 12 6zm0 5c-.83 0-1.5-.67-1.5-1.5S11.17 8 12 8s1.5.67 1.5 1.5S12.83 11 12 11z"
            />
          </g>
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root-116"
      />
    </a>
  </div>
  <div
    className="Component-container-3"
  >
    <a
      aria-current="page"
      href="/"
      onClick={[Function]}
      style={
        Object {
          "textDecoration": "none",
        }
      }
    >
      <img
        alt="Prizey"
        className="Component-logoImage-7"
        src="/logo.png"
      />
    </a>
    <div
      className="Component-caption-1"
    >
      <div
        className="Component-root-49"
      >
        <span
          className="Component-icon-45 Component-icon_easy-46"
        />
        <h6
          className="MuiTypography-root-50 MuiTypography-subtitle1-68"
        >
          EASY
        </h6>
      </div>
    </div>
    <div
      className="Component-children-2"
    >
      <h5
        className="MuiTypography-root-50 MuiTypography-h5-66 MuiTypography-alignCenter-73"
      >
        Trump mask
      </h5>
      <img
        alt="Trump mask"
        className="Component-productImage-86"
        src="/mock/trump-mask.png"
        style={Object {}}
      />
      <p
        className="MuiTypography-root-50 MuiTypography-body1-59 MuiTypography-alignCenter-73"
      >
        MSRP: $
      </p>
      <div
        afterCreate={[Function]}
        name="tickets"
        path="/ticket_transactions"
        render={[Function]}
      >
        New - 
        <div>
          render - 
          <button
            aria-label="Sell it Back"
            className="MuiButtonBase-root-16 MuiButton-root-90 MuiButton-contained-101 MuiButton-containedSecondary-103 MuiButton-raised-104 MuiButton-raisedSecondary-106 MuiButton-fullWidth-115"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-91"
            >
              <span>
                SELL IT BACK
              </span>
              <div
                className="Component-rightIcon-89"
              >
                <img
                  alt="diamond"
                  className="Component-icon-87"
                  src="/icons/diamond.png"
                />
                <span
                  className="Component-quantity-88"
                >
                  3
                </span>
              </div>
            </span>
            <span
              className="MuiTouchRipple-root-116"
            />
          </button>
        </div>
      </div>
      <button
        aria-label="I want it"
        className="MuiButtonBase-root-16 MuiButton-root-90 MuiButton-contained-101 MuiButton-containedPrimary-102 MuiButton-raised-104 MuiButton-raisedPrimary-105 MuiButton-fullWidth-115"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-91"
        >
          I WANT IT!
        </span>
        <span
          className="MuiTouchRipple-root-116"
        />
      </button>
    </div>
    <div
      className="Component-footer-4"
    >
      <a
        className="Component-footerLink-5"
        href="/terms_of_service"
        onClick={[Function]}
        rel="noopener noreferer"
        target="_blank"
      >
        <span
          className="MuiTypography-root-50 MuiTypography-caption-60 MuiTypography-colorTextSecondary-83 MuiTypography-alignCenter-73"
        >
          TERMS
        </span>
      </a>
      <a
        className="Component-footerLink-5"
        href="https://app.termly.io/document/privacy-policy/8192f107-a2dc-4548-bf83-97907c6d83c3"
        onClick={[Function]}
        rel="noopener noreferer"
        target="_blank"
      >
        <span
          className="MuiTypography-root-50 MuiTypography-caption-60 MuiTypography-colorTextSecondary-83 MuiTypography-alignCenter-73"
        >
          PRIVACY
        </span>
      </a>
      <a
        className="Component-footerLink-5"
        href="/faq"
        onClick={[Function]}
        rel="noopener noreferer"
        target="_blank"
      >
        <span
          className="MuiTypography-root-50 MuiTypography-caption-60 MuiTypography-colorTextSecondary-83 MuiTypography-alignCenter-73"
        >
          FAIRNESS / FAQ
        </span>
      </a>
      <a
        className="Component-footerLink-5"
        href="/all-prizes"
        onClick={[Function]}
        rel="noopener noreferer"
        target="_blank"
      >
        <span
          className="MuiTypography-root-50 MuiTypography-caption-60 MuiTypography-colorTextSecondary-83 MuiTypography-alignCenter-73"
        >
          ALL PRIZES
        </span>
      </a>
    </div>
  </div>
</div>
`;
